### PR TITLE
[QENG-2164] [do not merge] Allow for daemonizing clamps non-root user agents

### PR DESCRIPTION
**Pull request by rick**
:warning: **work in progress** :warning:

Have clamps support daemonizing the non-root user agents, based on a `daemonize` parameter.

/cc @doug-rosser @ericwilliamson @acidprime

### CODE REVIEW

File: manifests/users.pp [hunk 1]
```diff
@@ -40,22 +41,31 @@
...
 
-  if $metrics_server {
-    file { "/home/${user}/time-puppet-run.sh":
-      ensure => file,
-      content => "TIMEFORMAT=\"metrics.${::fqdn}.${user}.time %R `date +%s`\"; TIME=$( { time /opt/puppet/bin/puppet agent --onetime --no-daemonize > /dev/null; } 2>&1 ); echo \$TIME | nc ${metrics_server} ${metrics_port}",
+  if $daemonize {
+    # TODO: we need to capture run timings for publication to graphite
+
+    exec { "user ${user} daemon puppet agent":
+      command => "/usr/bin/sudo -u ${user} /opt/puppet/bin/puppet agent --daemonize >/dev/null 2>&1",
```
Comment (rick) [inline manifests/users.pp]: I'm not thrilled here.  The intent is "have the non-root user(s) run this version of the agent". It works using `sudo`.

There are two significant downsides:

 1. Um, `sudo`.
 2. `sudo` on CentOS will fail when there is no controlling terminal, unless one modifies `/etc/sudoers` ([as documented here](https://www.shell-tips.com/2014/09/08/sudo-sorry-you-must-have-a-tty-to-run-sudo/))

/cc @doug-rosser @ericwilliamson

**Comment by rick** [comment]
This "works", though I don't like the `sudo` bit (as mentioned in the code comment earlier), and, also, the daemonized agent isn't easy to measure. I presume we can use the reports data to get run times and shove that into metrics.  I don't know if we've already covered that ground elsewhere.  

/cc @doug-rosser who probably knows about the metrics bit
/cc @ericwilliamson for another set of eyes in general

File: manifests/users.pp (hunk 1 above)
Comment (acidprime) [inline manifests/users.pp]: @rick did you try using the `user` param of the exec type

[reference](https://docs.puppetlabs.com/references/latest/type.html#exec-attribute-user)

Comment (acidprime) [inline manifests/users.pp]: Based on the notes, you might have to manually specify HOME for the shell to work but that can be done with the `environment` param

[reference](https://docs.puppetlabs.com/references/latest/type.html#exec-attribute-environment)

Comment (rick) [inline manifests/users.pp]: @acidprime I did not try that (yet).  Thanks for suggesting it. My puppet-fu is still fairly weak, so always happy for the pointers.

**Comment by rick** [comment]
I think this is about ready to merge now.  Could I get another set of :eyes: here?

**Comment by ericwilliamson** [comment]
:+1:

**Pull request merged**